src/tasks/dump.ts

```diff
@@ -0,0 +1,567 @@
+  BigNumber,
+  constants,
+  ContractTransaction,
+  Signer,
+  utils,
+} from "ethers";
+import { task, types } from "hardhat/config";
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+
+import { estimateTradeAmount, getFee, placeOrder } from "../services/api";
+import {
+  BUY_ETH_ADDRESS,
+  domain,
+  Order,
+  OrderKind,
+  SigningScheme,
+  signOrder,
+  TypedDataDomain,
+} from "../ts";
+
+import {
+  getDeployedContract,
+  isSupportedNetwork,
+  SupportedNetwork,
+} from "./ts/deployment";
+import { promiseAllWithRateLimit } from "./ts/rate_limits";
+import { Align, displayTable } from "./ts/table";
+import {
+  isNativeToken,
+  nativeToken,
+  NativeToken,
+  tokenDetails,
+  TokenDetails,
+  WRAPPED_NATIVE_TOKEN_ADDRESS,
+  balanceOf,
+  transfer,
+  displayName,
+} from "./ts/tokens";
+import { formatTokenValue } from "./ts/value";
+
+const MAX_LATEST_BLOCK_DELAY_SECONDS = 2 * 60;
+const MAX_ORDER_VALIDITY_SECONDS = 24 * 3600;
+
+interface DumpInstruction {
+  token: TokenDetails;
+  amount: BigNumber;
+  approvedAmount: BigNumber;
+  receivedAmount: BigNumber;
+  balance: BigNumber;
+  fee: BigNumber;
+}
+
+interface DisplayDumpInstruction {
+  fromSymbol: string;
+  fromAddress: string;
+  balance: string;
+  needsAllowance: "yes" | "";
+  receivedAmount: string;
+  feePercent: string;
+}
+
+interface TransferToReceiver {
+  token: TokenDetails | NativeToken;
+  amount: BigNumber;
+}
+
+interface DumpInstructions {
+  instructions: DumpInstruction[];
+  toToken: TokenDetails | NativeToken;
+  // Amount of toToken that will be transfered to the receiver address with a
+  // standard erc20 transfer. This is only set if there is a receiver and the
+  // specified toToken is also a token to be dumped.
+  transferToReceiver?: TransferToReceiver;
+}
+
+interface GetDumpInstructionInput {
+  dumpedTokens: string[];
+  toTokenAddress: string | undefined; // undefined defaults to native token (e.g., ETH)
+  user: string;
+  allowanceManager: string;
+  maxFeePercent: number;
+  hasCustomReceiver: boolean;
+  hre: HardhatRuntimeEnvironment;
+  network: SupportedNetwork;
+}
+async function getDumpInstructions({
+  dumpedTokens: inputDumpedTokens,
+  toTokenAddress,
+  user,
+  allowanceManager,
+  maxFeePercent,
+  hasCustomReceiver,
+  hre,
+  network,
+}: GetDumpInstructionInput): Promise<DumpInstructions> {
+  // todo: support dumping ETH by wrapping them
+  if (inputDumpedTokens.includes(BUY_ETH_ADDRESS)) {
+    throw new Error(
+      `Dumping the native token is not supported. Remove the ETH flag address ${BUY_ETH_ADDRESS} from the list of tokens to dump.`,
+    );
+  }
+
+  const dumpedTokens = Array.from(new Set(inputDumpedTokens));
+  const toToken =
+    toTokenAddress === undefined || toTokenAddress === BUY_ETH_ADDRESS
+      ? nativeToken(hre)
+      : await tokenDetails(toTokenAddress, hre);
+
+  let transferToReceiver: TransferToReceiver | undefined = undefined;
+  const sameTokenIndex = dumpedTokens.findIndex(
```

Comment:
Maybe just filter all tokens that are not equal to make the logic a bit easier to read (I had to look up what splice does - it's not the most intuitive method)

Reply:
I still can't tell what this is for, maybe a comment to go along with it. If it's anything like Felix is hinting at I imagine we could simplify things by casting the list to a set (and back if necessary).